radicle-node/src/worker.rs

```diff
@@ -190,153 +189,42 @@ impl<G: Signer + Ecdh + 'static> Worker<G> {
-                                    // is still sending stuff, we simply keep reading until we
-                                    // get a `done` packet.
-                                    continue;
-                                }
-                                Err(_) => {
-                                    // If we get any other error, eg. a timeout, we abort.
-                                    log::error!(
-                                        target: "worker",
-                                        "Upload of {} to {} aborted: missing `done` packet from remote",
-                                        fetch.rid,
-                                        fetch.remote
-                                    );
-                                    return (
-                                        WireSession::from_split_io(stream_r, stream_w),
-                                        Err(FetchError::RemoteAbortedFetch),
-                                    );
-                                }
-                            }
-                        }
-                    }
-                }
+                // XXX: two fetches are expected to happen, one for
+                // the `rad` refs, followed by the refs listed in
+                // signed refs.
+                let result = self._upload_pack(fetch, &mut pktline_r, &mut stream_w);
+                let result =
+                    result.and_then(|_| self._upload_pack(fetch, &mut pktline_r, &mut stream_w));
+                (WireSession::from_split_io(stream_r, stream_w), result)
             }
         }
     }
 
     fn fetch(
         &self,
-        rid: Id,
+        staging: fetch::StagingRad,
         namespaces: &Namespaces,
-        tunnel: &mut Tunnel<WireSession<G>>,
-    ) -> Result<Vec<RefUpdate>, FetchError> {
-        let repo = match self.storage.repository_mut(rid) {
-            Ok(r) => Ok(r),
-            Err(e) if e.is_not_found() => self.storage.create(rid),
-            Err(e) => Err(e),
-        }?;
-        let tunnel_addr = tunnel.local_addr()?;
-        let mut cmd = process::Command::new("git");
-        cmd.current_dir(repo.path())
-            .env_clear()
-            .envs(env::vars().filter(|(k, _)| k == "PATH" || k.starts_with("GIT_TRACE")))
-            .envs(git::env::GIT_DEFAULT_CONFIG)
-            .args(["-c", "protocol.version=2"])
-            .arg("fetch")
-            .arg("--verbose");
+        session: WireSession<G>,
+    ) -> (WireSession<G>, Result<Vec<RefUpdate>, FetchError>) {
+        let (session, res) =
+            self.tunnel_fetch(&staging.repo, staging.refspecs(namespaces.clone()), session);
+
+        let staging = match res.and_then(|()| {
+            staging
+                .into_staging_refs(namespaces.clone())
+                .map_err(FetchError::from)
+        }) {
+            Ok(staging) => staging,
+            Err(e) => return (session, Err(e)),
+        };
```

Comment:
This -might- be easier to grok if the and_then was split into a second `let .. match(.)`.  It looks like both would be returning on error.

Reply:
Ya, I'm avoiding having to repeat matching on the errors. Is it's current form confusing?